### DISCUSSION TITLE
Smart Validation not working React datepicker inside hook form

### DISCUSSION THREAD

**Discussion by gboggar** [react-hook-form/react-hook-form#2345] [Q&A]
I am using Datepicker inside hook form. My intention is to implement smart validation before save. isDirty property of the form is true when I tried to change the date but the same date selected again. The form is thinking the value changed but in reality value is the same as the original. 

Following code I am using:
<Controller
				as={
					<DatePicker
						dateFormat={props.dateFormat} 
						showTimeSelect={props.showTimeSelect}
						todayButton="Today"
						ref={(r) => dateRef = r}
	
						dropdownMode="select"
						placeholderText={props.placeholder}  
						shouldCloseOnSelect
						isClearable
					/>
				}
				control={props.control}
				name={props.name}
				valueName="selected" 
				onChangeRaw={(e) => e.preventDefault()}
				onChange={([selected]) => props.onChangeFunc(selected)}
				className="form-control"
				rules={props.validationRules}
			/>

Any help from the community greatly appreciated.

**Comment by bluebill1049** [comment]
take a look the doc, there is an example with codesandbox with datetimepicker

**Reply by gboggar** [reply]
Can you please direct me to the sample that you are mentioning.  The code sample that I found is not working the way it supposed to be. once the date is touched it remains dirty even after selecting the same value it was first initialized. sorry for the trouble.

**Reply by bluebill1049** [reply]
is this the one?https://codesandbox.io/s/react-hook-form-v6-controller-qsd8r

**Reply by gboggar** [reply]
The code sample is working for react date picker and other controls. Maybe I am not clear on my question. I want formState.dirty to be false if I touch the datepicker to make selection but  selected the date back to original default selection when loaded. In the Sample formState.dirty remains true even after putting date selection back to original. Maybe hook-forms not handling that scenario yet. Thanks for your quick response.

**Reply by bluebill1049** [reply]
because there is no codesandbox, i have to guess: you need to set your mode to 'onChange'

**Reply by QuicklynX19** [reply]
This is not the case 

> you need to set your mode to 'onChange'

I have come across this issue as well, where when we want to update a react-datepicker field, once selecting the defaultValue date, it will still mark the form dirty (dirtyFields has the date input inside). I would guess that the issue here is that two Date objects are being compared and as stated [here](https://stackoverflow.com/questions/7606798/javascript-date-object-comparison) two Date objects are not the same, no matter if the dates are. Would you mind taking a look at this?

**Steps to reproduce:**

1. Visit [https://codesandbox.io/s/react-hook-form-v6-controller-qsd8r](https://codesandbox.io/s/react-hook-form-v6-controller-qsd8r)
2. Observe the `formState` field (either console.log it or use react developer tools)
3. Change the React-datepicker date to a different defaultValue
4. Change the React-datepicker date back to defaultValue

**Bug**

The form will still mark the date as dirty, even though the initial value was selected.
